src/asl.ts

```diff
@@ -430,19 +430,63 @@ export const isWaitState = (state: State): state is Wait => {
   return state.Type === "Wait";
 };
 
+type EvalExprHandler<Output extends ASLGraph.Output = ASLGraph.Output> = (
+  output: Output,
+  context: EvalExprContext
+) => ASLGraph.NodeResults;
+
 export interface EvalExprContext {
   /**
-   * returns a possibly mutated output variable as json path.
-   *
-   * If the output was a {@link ASLGraph.LiteralValue}, a new state will be added that turns the literal into a json path.
-   * If the output was a {@link ASLGraph.JsonPath}, the output is returned.
+   * Callback provided to inject additional states into the graph.
+   * The state will be joined (@see ASLGraph.joinSubStates ) with the previous and next states in the order received.
    */
-  normalizeOutputToJsonPath: () => ASLGraph.JsonPath;
+  addState: (state: ASLGraph.NodeState | ASLGraph.SubState) => void;
+}
+
+type EvalContextHandler = (context: EvalContextContext) => ASLGraph.NodeResults;
+
+export interface EvalContextContext {
   /**
    * Callback provided to inject additional states into the graph.
    * The state will be joined (@see ASLGraph.joinSubStates ) with the previous and next states in the order received.
    */
   addState: (state: ASLGraph.NodeState | ASLGraph.SubState) => void;
+  /**
+   * Evaluates a single expression and returns the {@link ASLGraph.Output}.
+   *
+   * Any require states will be merged in with the output.
+   *
+   * This method is the same as {@link ASL.evalExpr}, but can be used multiple times within a single {@link ASL.evalContext}.
+   */
+  evalExpr: (expr: Expr, allowUndefined?: boolean) => ASLGraph.Output;
+  /**
+   * Evaluates a single expression and returns the {@link ASLGraph.Output}.
+   *
+   * Any require states will be merged in with the output.
+   *
+   * * If the output was a {@link ASLGraph.LiteralValue}, a new state will be added that turns the literal into a {@link ASL.JsonPath}.
+   * * If the output was a {@link ASLGraph.ConditionOutput}, a new {@link Choice} state will turn the conditional into a boolean
+   *
+   * This method is the same as {@link ASL.evalExpr}, but can be used multiple times within a single {@link ASL.evalContext}.
+   */
+  evalExprToJsonPath: (
+    expr: Expr,
+    allowUndefined?: boolean
+  ) => ASLGraph.JsonPath;
+  /**
+   * Evaluates a single expression and returns the {@link ASLGraph.JsonPath} or {@link ASLGraph.LiteralValue}.
+   *
+   * Any require states will be merged in with the output.
+   *
+   * If the output was a {@link ASLGraph.ConditionOutput}, a new {@link Choice} state will turn the conditional into a boolean
+   * and return a {@link ASLGraph.JsonPath}.
+   *
+   * This method is the same as {@link ASL.evalExpr}, but can be used multiple times within a single {@link ASL.evalContext}.
+   */
+  evalExprToJsonPathOrLiteral: (
```

Comment:
How are these two different?

Reply:
first one normalizes any output to a `JsonPath`, second one normalizes any output to a JsonPath or Literal value (aka, turns a condition into a JsonPath by adding a conditional state `heap = (cond) ? true : false`).